File: src/detect/variant.ts
import type { ThemeMode } from "@/lib/types";
import { AUTO_APPLY_MODE_CONFIDENCE, inferThemeMode, resolveModeHint } from "@/lib/mode";
import type { VariantModeResult } from "./types";

const CONFIDENCE = {
  FULL: 1.0,
  NONE: 0,
} as const;

const SOURCE = {
  PATTERN: "pattern",
  HINT: "hint",
  UNKNOWN: "unknown",
} as const;

export function detectVariantModeFromName(variantName: string): ThemeMode | undefined {
  const inference = inferThemeMode(variantName);
  if (!inference) return undefined;
  if (inference.confidence < AUTO_APPLY_MODE_CONFIDENCE) return undefined;
  return inference.mode;
}

export type VariantInput = {
  name: string;
  colorscheme?: string;
  mode?: ThemeMode | undefined;
  meta?: { strategy?: unknown };
};

export function detectVariantModesFromNames(
  variants: VariantInput[] | undefined,
): VariantModeResult[] {
  if (!variants || variants.length === 0) return [];

  return variants.map((variant): VariantModeResult => {
    const inference = inferThemeMode(variant.name);
    if (!inference) {
      return { name: variant.name, confidence: CONFIDENCE.NONE, source: SOURCE.UNKNOWN };
    }

    if (inference.confidence < AUTO_APPLY_MODE_CONFIDENCE) {
      return {
        name: variant.name,
        confidence: inference.confidence,
        source: SOURCE.UNKNOWN,
        reason: `Low-confidence pattern match: ${inference.reason}`,
      };
    }

    return {
      name: variant.name,
      detectedMode: inference.mode,
      confidence: inference.confidence,
      source: SOURCE.PATTERN,
      reason: inference.reason,
    };
  });
}

export function applyVariantHints(
  results: VariantModeResult[],
  hints: Record<string, ThemeMode>,
): VariantModeResult[] {
  return results.map((result): VariantModeResult => {
    const hint = resolveModeHint(result.name, hints);
    if (!hint) return result;

    const reason = hint.normalizedMatch
      ? `Manual hint override (normalized match: ${hint.matchedKey})`
      : "Manual hint override";

    return {
      ...result,
      detectedMode: hint.mode,
      confidence: CONFIDENCE.FULL,
      source: SOURCE.HINT,
      reason,
    };
  });
}

File: internal/detect/variant.go
package detect

import (
	"fmt"

	"github.com/raulcorreia7/theme-browser-registry/internal/theme"
)

// Variant coverage sources (SOURCE constants in src/detect/variant.ts).
const (
	SourcePattern = "pattern"
	SourceHint    = "hint"
	SourceReadme  = "readme"
	SourceUnknown = "unknown"
)

// DetectVariantModesFromNames ports detectVariantModesFromNames: pattern
// detection over variant names via theme.InferThemeMode. Only the name is
// consulted, exactly like the TypeScript implementation.
func DetectVariantModesFromNames(variants []theme.Variant) []VariantModeResult {
	if len(variants) == 0 {
		return []VariantModeResult{}
	}

	results := make([]VariantModeResult, 0, len(variants))
	for _, v := range variants {
		inf := theme.InferThemeMode(v.Name)
		if inf == nil {
			results = append(results, VariantModeResult{
				Name:       v.Name,
				Confidence: 0,
				Source:     SourceUnknown,
			})
			continue
		}

		if inf.Confidence < HighConfidenceThreshold {
			results = append(results, VariantModeResult{
				Name:       v.Name,
				Confidence: inf.Confidence,
				Source:     SourceUnknown,
				Reason:     fmt.Sprintf("Low-confidence pattern match: %s", inf.Reason),
			})
			continue
		}

		results = append(results, VariantModeResult{
			Name:         v.Name,
			DetectedMode: inf.Mode,
			Confidence:   inf.Confidence,
			Source:       SourcePattern,
			Reason:       inf.Reason,
		})
	}
	return results
}

// ApplyVariantHints ports applyVariantHints: manual hints win over pattern
// matches with confidence 1.0.
func ApplyVariantHints(results []VariantModeResult, hints map[string]theme.Mode) []VariantModeResult {
	out := make([]VariantModeResult, len(results))
	copy(out, results)
	for i := range out {
		hint := theme.ResolveModeHint(out[i].Name, hints)
		if hint == nil {
			continue
		}
		reason := "Manual hint override"
		if hint.NormalizedMatch {
			reason = fmt.Sprintf("Manual hint override (normalized match: %s)", hint.MatchedKey)
		}
		out[i].DetectedMode = hint.Mode
		out[i].Confidence = 1.0
		out[i].Source = SourceHint
		out[i].Reason = reason
	}
	return out
}
